File: Diagrammatize.Design.Pattern/code/python/Interpreter/Sample/CommandNode.py

from Node import Node

# <command> ::= <repeat command> | <primitive command>
class CommandNode(Node):
    def __init__(self):
        super().__init__()
        self.node = None

    def parse(self, context):
        from RepeatCommandNode import RepeatCommandNode
        from PrimitiveCommandNode import PrimitiveCommandNode
        if context.currentToken() == "repeat":
            self.node = RepeatCommandNode()
            self.node.parse(context)
        else:
            self.node = PrimitiveCommandNode()
            self.node.parse(context)

    def __str__(self):
        return str(self.node)

File: Diagrammatize.Design.Patterns/code/cxx/Interpreter/Sample/CommandNode.cpp
#include "CommandNode.hpp"
#include "PrimitiveCommandNode.hpp"
#include "RepeatCommandNode.hpp"

CommandNode::~CommandNode() {}

void CommandNode::parse(Context& context) {
    if (context.currentToken() == "repeat") {
        node_.reset(new RepeatCommandNode());
        node_->parse(context);
    } else {
        node_.reset(new PrimitiveCommandNode());
        node_->parse(context);
    }
}

std::string CommandNode::toString() {
    return node_->toString();
}
